### PR TITLE
Label print count: a count on all four print dialogs, and repair the one that never printed (#86)

### LINKED ISSUE #86: Label Printing Dialogs Count / Quantity
Label printing (both when adding/editing a single material inventory item as well as when bulk printing labels for multiple items) should support a quantity/count of labels to print. The UI input field for this should default to 1. When bulk printing labels for multiple items, the each item's label should be printed the specified quantity/count number of times.

### PR DESCRIPTION
**Pull request by jantman**
Closes #86.

Adds a **label count** — a whole number 1–99, defaulting to 1 — to every dialog that prints an inventory item's label, and repairs the one of those four that has never been able to print at all.

Spec, plan and tasks are in `specs/013-label-print-count/` (specified in eeb9762, designed in 9c88143).

## The mechanism was already in the tree, and unreachable

`generate_and_print_label()` took a `num_copies` parameter that builds `[image] * n` for one `lp` job, and nothing ever passed it a value, because `print_label_for_ja_id()` didn't thread it through. So the backend here is one integer down two calls plus route validation. No new printing path, no dependency, no schema change, no migration.

`label_count` is **optional and absent means 1**, which is what kept every existing caller valid and let the four dialogs convert one at a time. The route rejects `bool` explicitly — `isinstance(True, int)` is `True` in Python, so a naive int check would accept `true` and print one label for it.

## One helper, four dialogs

`app/static/js/label-count.js` owns the 1–99 bounds and the error wording for all four call sites, so they can't drift. SC-007 makes their agreement a requirement, and the post-bulk-Add dialog is the standing cautionary tale of a fourth copy going its own way. The gate is the helper and the route, not the input's `min`/`max`: every print button is `type="button"`, so browser constraint validation never fires, and a validation bubble isn't observable from an e2e test.

## The FR-012 repair

The dialog offered after a bulk Add Item creation posted `{ja_id, label_size}` to an endpoint that requires `label_type` and has no field called `label_size` — so **every press of "Print All Labels" has always returned 400**. Its three invented sizes are replaced by the six real types from `/api/labels/types`, Print All stays disabled until one is picked, the payload is right, and it now surfaces `data.error` instead of discarding it for `response.statusText`.

## Reporting

| Surface | At count 1 | Above 1 |
|---|---|---|
| Single-item alert | unchanged, verbatim | `3 labels printed successfully for JA123456` |
| Bulk progress line | unchanged, verbatim | `Printing 1 of 3: JA…  (2 labels)` |
| Bulk summary | **reworded** — `Complete: 3 labels for 3 items, 0 failed` | same shape |

The summary changes at every count because FR-008 requires both numbers, and the "indistinguishable at 1" edge case protects against *friction* rather than phrasing — nothing was added to the count-1 path. A failed item contributes **zero** labels, never a partial figure: one item's copies are a single `lp` job with one exit code, so under-claiming is the only safe direction.

## Two things worth a reviewer's attention

**1. The bulk error div moved out of the progress region.** It was nested inside a container that's `d-none` until printing starts. A refused label count is reported *before* any run begins, so the message would have rendered where nobody could see it and FR-005 couldn't have held. Now a sibling, in both templates.

**2. T024 (screenshot regeneration) ran and produced nothing to commit.** The task's premise doesn't hold, and the reasoning is recorded against T024 in `tasks.md`:
- No screenshot of any label printing dialog exists on disk.
- `screenshot_config.yaml:85` names a `test_screenshot_bulk_label_printing` that isn't defined anywhere.
- The test that *would* capture that modal (`test_screenshot_label_printing`) swallows its failure in a `try/except`, so it has never produced an image.
- Regeneration is non-deterministic — two runs on identical code produced different bytes for the same six *unrelated* PNGs.

Committing that would have added eight randomly-differing binaries and no information, so it was reverted. Fixing the orphaned config entry and the swallowed failure is a pre-existing gap, out of scope here.

## Verification

- `nox -s tests` — **1295 passed** (22 new; baseline was 1273)
- `nox -s e2e` — **498 passed**, full suite, 11m23s, **zero fixed waits added**
- Route curled over real HTTP on a throwaway SQLite store: all eight boundary cases match the contract, including `label_count: 500` → `label_count must be between 1 and 99` and absent → today's unchanged message
- `nox -s lint` red only on pre-existing E501s, unchanged from baseline

No test reaches `LpPrinter.print_images()` — the test-mode short-circuit stays first in the function.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01QDUzzRSuimKp4hy1oVwc5K

### CODE REVIEW

File: app/static/js/inventory-add.js [hunk 1]
```diff
@@ -874,6 +930,16 @@ class InventoryAddForm {
         const doneBtn = document.getElementById('bulk-print-done-btn');
         const skipBtn = document.getElementById('bulk-print-skip');
 
+        // Read the count before anything is printed -- a refused count must
+        // leave the dialog untouched and print nothing at all.
+        const countResult = window.readLabelCount('bulk-label-count');
```
[the remark this answers is not in the document]
Comment (jantman) [reply app/static/js/inventory-add.js:951]: Confirmed and fixed in a5190d3 — you and Copilot both caught this, and the diagnosis is exactly right.

The only clear was in `showBulkLabelPrintingModal()`, which runs on open, not on a click. So `0` → refused → correct to `3` → Print All left the warning sitting directly above `Complete: 6 labels for 2 items, 0 failed`. A successful run that reads as a failed one, and self-inflicted: it became reachable precisely *because* this PR moved the div out of the `d-none` progress container to make refusals visible at all.

Took the suggestion as written, in both dialogs. Added a regression test per dialog (`test_post_creation_corrected_count_clears_the_earlier_warning`, `test_bulk_corrected_count_clears_the_earlier_warning`): refuse, correct, print, then assert the warning is gone — ordered so the completion line establishes the region first, otherwise the negative assertion would pass against a dialog that had not finished rendering.

File: app/static/js/inventory-list.js [hunk 2]
```diff
@@ -192,6 +192,16 @@ class InventoryListManager {
         const doneBtn = document.getElementById('list-bulk-print-done-btn');
         const cancelBtn = document.getElementById('list-bulk-print-cancel');
 
+        // Read the count before anything is printed -- a refused count must
+        // leave the dialog untouched and print nothing at all.
+        const countResult = window.readLabelCount('list-bulk-label-count');
```
[the remark this answers is not in the document]
Comment (jantman) [reply app/static/js/inventory-list.js:204]: Fixed in a5190d3, same change as the sibling — thanks for pointing at both call sites rather than just the one.

`resetBulkPrintModal()` runs on open and close, never on a Print All click, so the warning survived the retry. Took the suggestion as written and added `test_bulk_corrected_count_clears_the_earlier_warning` to hold it.

File: app/static/js/inventory-list.js [hunk 3]
```diff
@@ -192,6 +192,16 @@ class InventoryListManager {
...
+        if (!countResult.ok) {
+            errorsDiv.classList.remove('d-none');
+            errorsDiv.innerHTML = `<strong>Warning:</strong> ${countResult.error}`;
+            return;
+        }
+        const labelCount = countResult.value;
```
[the remark this answers is not in the document]
Comment (jantman) [reply app/static/js/inventory-list.js:210]: Correct, and fixed in a5190d3.

`resetBulkPrintModal()` only runs when the dialog opens and closes, so nothing cleared the region on a Print All click. `errorsDiv` is now hidden and emptied at the top of `printAllLabels()`, before the count is validated, so a corrected retry starts from a clean dialog. Regression test: `test_bulk_corrected_count_clears_the_earlier_warning`.

Worth noting this was reachable only because the PR moved the div out of the `d-none` progress container — before that a refusal rendered somewhere invisible, which was the worse bug of the two.

File: app/static/js/inventory-add.js [hunk 4]
```diff
@@ -874,6 +930,16 @@ class InventoryAddForm {
...
+        if (!countResult.ok) {
+            errorsDiv.classList.remove('d-none');
+            errorsDiv.innerHTML = `<strong>Warning:</strong> ${countResult.error}`;
+            return;
+        }
+        const labelCount = countResult.value;
```
[the remark this answers is not in the document]
Comment (jantman) [reply app/static/js/inventory-add.js:957]: Correct, and fixed in a5190d3 — same change as the list dialog.

Here the clear lived in `showBulkLabelPrintingModal()`, so it only ran on open. `errorsDiv` is now hidden and emptied at the top of `printAllLabels()` before validation. Regression test: `test_post_creation_corrected_count_clears_the_earlier_warning`.

File: app/static/js/inventory-list.js [hunk 5]
```diff
@@ -245,16 +259,21 @@ class InventoryListManager {
...
 
-        // Update final status
-        statusSpan.textContent = `Complete: ${successCount} printed, ${failureCount} failed`;
+        // Update final status. A failed item contributes 0 labels rather than a
+        // partial figure -- one item's copies are one lp job with one exit code,
+        // so the total must never claim more labels than actually emerged.
+        const labelsPrinted = successCount * labelCount;
+        statusSpan.textContent =
+            `Complete: ${labelsPrinted} labels for ${successCount + failureCount} items, ` +
+            `${failureCount} failed`;
```
[the remark this answers is not in the document]
Comment (jantman) [reply app/static/js/inventory-list.js:277]: Agreed, fixed in a5190d3. One selected item at a count of 1 now reads `Complete: 1 label for 1 item, 0 failed` — I pluralized *both* nouns, since `labels` hits 1 in the same case and `1 labels for 1 items` was the worse of the two readings.

Flagging that this is a deliberate deviation from a specified string, so it is a judgment call a reviewer may want to overrule. `research.md` Decision 5 and `plan.md` both pin the format as `Complete: {labels} labels for {items} items, {failed} failed`. I read `1 items` as an unconsidered consequence of a format written against a three-item illustration rather than an intended outcome — both numbers are still reported at every count, so FR-008 and the reasoning behind Decision 5 are untouched.

`research.md` and `plan.md` are left as written; per this repo`&#39;`s CLAUDE.md those are the frozen record of what was specified, not a description of what shipped. The deviation is recorded against T014 in `tasks.md` instead.

E2E expectations updated (one assertion actually changed: `4 labels for 1 items` → `4 labels for 1 item`), plus `test_bulk_summary_uses_singular_nouns_for_one` covering both singular branches.

File: app/static/js/inventory-add.js [hunk 6]
```diff
@@ -924,15 +998,20 @@ class InventoryAddForm {
             `;
         }
 
-        statusSpan.textContent = `Complete: ${successCount} printed, ${failureCount} failed`;
+        // A failed item contributes 0 labels rather than a partial figure, so
+        // the total never claims more labels than actually emerged.
+        const labelsPrinted = successCount * labelCount;
+        statusSpan.textContent =
+            `Complete: ${labelsPrinted} labels for ${successCount + failureCount} items, ` +
+            `${failureCount} failed`;
```
[the remark this answers is not in the document]
Comment (jantman) [reply app/static/js/inventory-add.js:1024]: Pluralization applied here too in a5190d3, so both dialogs report identically (FR-013).

One thing worth recording: the singular branch is **unreachable on this dialog**. It only opens for a quantity of 2 or more — a quantity of 1 takes the ordinary single-item path — so `1 item` cannot occur here. Rather than write a test that could not exercise it, I covered the singular on the list page`&#39;`s dialog, where a single item can be selected, and left a comment at the end of `test_bulk_creation.py` saying why the case is absent so the gap does not later read as an oversight.

The reasoning about deviating from the spec`&#39;`s literal format string is in my reply on the list dialog.

File: app/static/js/inventory-add.js [hunk 7]
```diff
@@ -806,6 +806,43 @@ class InventoryAddForm {
...
+
+            data.label_types.forEach(labelType => {
+                const option = document.createElement('option');
+                option.value = labelType;
+                option.textContent = labelType;
+                select.appendChild(option);
+            });
+        } catch (error) {
+            console.error('Error loading label types:', error);
+        }
```
[the remark this answers is not in the document]
Comment (jantman) [reply app/static/js/inventory-add.js:846]: Good catch, fixed in a5190d3.

This one was an inconsistency I introduced: `loadBulkLabelTypes()` was written to mirror `loadLabelTypes()` in `inventory-list.js`, and that one *does* surface the failure. Mine dropped it and only logged, so the outcome was an empty select plus a permanently disabled Print button and nothing explaining either.

It now writes into the dialog`&#39;`s error region — which this PR made usable outside the progress container — rather than an `alert()`, since the region is right there and is assertable from a test. The message says the items were created and can be printed from the inventory list, because that is the actionable part at that moment: the creation already succeeded and only the printing offer is broken.

No test for it: forcing the failure means intercepting `/api/labels/types` to fail, which is a fair bit of scaffolding for a path whose whole behavior is one string in one div. Happy to add it if you would rather have it covered.

**Comment by jantman** [comment]
Review addressed in a5190d3. All seven inline comments were real; none were dismissed.

**1. Stale count warning survived a corrected retry** (4 comments, both reviewers, both dialogs). The only clear lived in the modal reset, which runs on open/close and never on a Print All click — so refusing a count, correcting it, and printing left the warning directly above a *successful* completion line. Self-inflicted: it became reachable because this PR moved the error div out of the `d-none` progress container to make refusals visible at all. Cleared at the top of `printAllLabels()` in both dialogs, one regression test each.

**2. `1 items`** (2 comments). Pluralized both nouns — `Complete: 1 label for 1 item, 0 failed`. This is a **deliberate deviation from a specified string**, so it is the one change here a reviewer may want to overrule: `research.md` Decision 5 pins the format as `{labels} labels for {items} items`. I read `1 items` as an unconsidered consequence of a format written against a three-item illustration; both numbers are still reported at every count, so FR-008 is untouched. `research.md`/`plan.md` left as written — frozen record — with the deviation logged against T014 in `tasks.md`.

**3. Silent label-type load failure** (1 comment). An inconsistency I introduced: this function was written to mirror `loadLabelTypes()` in `inventory-list.js`, which *does* surface the failure. Now writes to the dialog error region, saying the items were created and can be printed from the inventory list — the actionable part, since only the printing offer is broken.

**Verification:** `nox -s tests` 1295 passed; `nox -s e2e` **501 passed**, up from 498 (three new regression tests), working tree clean afterwards.

One aside on the 📸 screenshot reminder above: it says *"Regeneration is not reproducible today — rendered timestamps and form validation state vary between runs, so expect some files to change without a real UI difference."* That independently confirms what I found running T024 and documented in the PR description — I measured six unrelated PNGs differing between two runs on identical code, and reverted the churn rather than commit it.
